src/main/java/frc/robot/GenericPID/GenericPID.java

```diff
@@ -0,0 +1,114 @@
+package frc.robot.GenericPID;
+
+import java.awt.Color;
+import frc.robot.GenericPID.Approximations.*;
+import frc.robot.GenericPID.Testing.*;
+
+public class GenericPID {
+    //a basic pid controller class, that lets the motor handle tracking position, velocity, and max of these. 
+    //for this functionality, extend this class (could override controlEffect)
+    //the max effective target velocity should be implicitly in line with the tuning, and the motor should cap itself. This object is quite basic.
+    //the motor should listen to pid after each control effect, and keep track of its own time value too (this can be a real life time value)
+    //graph of how the pid time should update
+    //t .          .          .
+    //   |________| |________|
+    //    motor i0   motor i1    (you handle)
+    //  di0  di1        di2      (internal) 
+    //   i0        i1       i2   (internal)
+    //   p0        p1       p2   (internal)
+    
+    //todo: maybe genericPID Profile that can use any derivative sorta function? it is generic, but would be kinda useless
+
+    private static final boolean debug = Debug.debug_GenericPID;
+
+
+
+    private PIDConfig conf;
+    private double t;
+    private ApproximateDerivative dedt;
+    private ApproximateIntegral E;
+
+    public static void test() {
+        PIDConfig c = new PIDConfig();
+        c.kP = 1.7;
+        c.kI = 0.0001;
+        c.kD = 0.06;
+        GenericPID p = new GenericPID(c);
+        ArtificialMotor m = new ArtificialMotor(2, 0.1, 0, 2, 0, 0, 0);
+        double dt = 0.001;
+        double target = 2;
+        // Graph g = new Graph(new Graph.GraphConfig.Builder.y2(100).build());
+        Graph.GraphConfig gc = new Graph.GraphConfig();
+        gc.x1 = 0;
+        gc.x2 = 20;
+        gc.y2 = 3;
+        gc.y1 = -3;
+        Graph g = new Graph(gc);
+        g.addPlot(Color.BLUE);
+        class f implements DoubleFunction {
+            public double eval(double x) {
+                //first the control effect is updated (including with the derivative of currrent to last),
+                //then the motor is directed towards this velocity, and the motor updates its position and known time.
+                //then the pid updates its known time, should be in sync with the motor's known time
+                //see graph atop this class
+                if (x == 0) {
+                    m.sustain(p.firstEffect(target, m.p()), dt);
+                    p.next_t(dt);
+                }
+                m.sustain(p.controlEffect(target, m.p(), dt), dt); 
```

Comment:
just to summarize, you're getting the position of the motor every time this method is polled by putting the motor in a go/slow/active state, depending on the magnitude of the controller effect (kP*prop+kI*int+kD*der).

Reply:
yes